File: backend/routes/uploads.py
from fastapi import APIRouter, UploadFile, File, Form

from backend.models.upload import UploadFilterResult, UploadResponse

router = APIRouter(prefix="/api/uploads", tags=["uploads"])

ALLOWED_CONTENT_TYPES: dict[str, str] = {
    "application/pdf": "text",
    "image/jpeg": "image",
    "image/png": "image",
    "image/webp": "image",
    "text/csv": "text",
    "text/plain": "text",
    "application/vnd.ms-excel": "text",
    "video/mp4": "video",
    "video/quicktime": "video",
    "video/webm": "video",
}


def _detect_file_type(content_type: str, filename: str) -> str:
    if content_type in ALLOWED_CONTENT_TYPES:
        return ALLOWED_CONTENT_TYPES[content_type]
    ext = filename.rsplit(".", 1)[-1].lower() if "." in filename else ""
    if ext == "pdf":
        return "text"
    if ext in ("jpg", "jpeg", "png", "webp"):
        return "image"
    if ext in ("csv", "txt"):
        return "text"
    if ext in ("mp4", "mov", "webm"):
        return "video"
    return "unknown"


async def _ai_filter(file: UploadFile, file_type: str, contents: bytes) -> UploadFilterResult:
    """Stub — accepts every file. Replace with real AI filtering later."""
    return UploadFilterResult(
        filename=file.filename or "unnamed",
        file_type=file_type,  # type: ignore[arg-type]
        size=len(contents),
        accepted=True,
        reason="Passed (AI filter not yet active)",
        ai_tags=[],
        ai_confidence=None,
    )


@router.post("", response_model=UploadResponse)
async def upload_files(
    files: list[UploadFile] = File(...),
    program_id: str = Form(...),
):
    results: list[UploadFilterResult] = []

    for f in files:
        contents = await f.read()
        file_type = _detect_file_type(f.content_type or "", f.filename or "")
        result = await _ai_filter(f, file_type, contents)
        results.append(result)

    accepted = sum(1 for r in results if r.accepted)
    return UploadResponse(
        total_files=len(results),
        accepted=accepted,
        rejected=len(results) - accepted,
        program_id=program_id,
        results=results,
    )

File: lib/api/uploads.ts
const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000"

export interface QualityWarning {
  check: string
  message: string
}

export interface QualityScanResult {
  score: number
  passed: boolean
  reason: string
  warnings: QualityWarning[]
}

export interface UploadFilterResult {
  filename: string
  fileType: "image" | "text" | "video" | "unknown"
  size: number
  accepted: boolean
  reason: string | null
  url: string | null
  detectedLabel: string | null
  quality: QualityScanResult
  aiTags: string[]
  aiConfidence: number | null
}

export interface UploadResponse {
  totalFiles: number
  accepted: number
  rejected: number
  programId: string
  results: UploadFilterResult[]
}

export async function uploadFiles(
  files: File[],
  programId: string
): Promise<UploadResponse> {
  const formData = new FormData()
  for (const file of files) {
    formData.append("files", file)
  }
  formData.append("program_id", programId)

  const res = await fetch(`${API_BASE}/api/uploads`, {
    method: "POST",
    body: formData,
  })
  if (!res.ok) throw new Error(`Upload failed: ${res.status}`)

  const data = await res.json()
  return {
    totalFiles: data.total_files,
    accepted: data.accepted,
    rejected: data.rejected,
    programId: data.program_id,
    results: data.results.map((r: Record<string, unknown>) => ({
      filename: r.filename,
      fileType: r.file_type,
      size: r.size,
      accepted: r.accepted,
      reason: r.reason,
      url: r.url,
      detectedLabel: r.detected_label,
      quality: {
        score: (r.quality as Record<string, unknown>).score as number,
        passed: (r.quality as Record<string, unknown>).passed as boolean,
        reason: (r.quality as Record<string, unknown>).reason as string,
        warnings: ((r.quality as Record<string, unknown>).warnings as QualityWarning[]) ?? [],
      },
      aiTags: r.ai_tags,
      aiConfidence: r.ai_confidence,
    })),
  }
}
